### ISSUE TITLE
[ENH] Add CMake module for compiling Cython code

### ISSUE THREAD

**Issue by vyasr** [cython/cython#5486]
### Is your feature request related to a problem? Please describe.

Currently the recommended method for building Cython modules is [using setuptools via setup.py](https://cython.readthedocs.io/en/latest/src/quickstart/build.html). While this is still completely valid, the general trend of the Python packaging ecosystem is to move away from this pattern. This is true in multiple respects:
- [setuptools is migrating away from arbitrary code in setup.py to a static setup.cfg config file](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#transitioning-from-setup-py-to-setup-cfg)
- [PEP 517 specifies how to define build backends other than setuptools](https://peps.python.org/pep-0517/)
- Support for more complex builds involving Cython and extensive C/C++ libraries has dramatically improved over the years with the advent of tools like [`meson-python`](https://github.com/mesonbuild/meson-python) and [`scikit-build`](https://github.com/scikit-build/scikit-build/)

I do not know if Cython has any plans for supporting setup.cfg files (I wasn't able to find anything in the issue tracker), nor do I have enough knowledge of Cython internals to propose a plan for how Cython could be integrated into a setup.cfg-based build. For projects leveraging tools like meson or CMake, though, it there is a fairly should be possible to chart a reasonable path forward here.

### Describe the solution you'd like.

The general pattern for Python builds using an underlying build system is to use `cython` to transpile Cython code into C/C++ and then to manage compilation of the resulting source file using the build generator (as opposed to using `cythonize`) since that allows users to leverage the full features of the build system to control the compilation. I do not personally make use of Meson, but I do have experience with CMake and scikit-build. To help facilitate the transition for Cython users using more extensive CMake-based builds, I would like for Cython to provide a CMake module that (at minimum) defines a function that can be used to run the `cython` CLI and generate code with it. This function would provide an interface to all the command line flags for the `cython` CLI and handle any necessary preprocessing. The resulting module could then be used by e.g. `scikit-build` users or those manually handling CMake-based compilation.

### Describe alternatives you've considered.

[`scikit-build` currently provides a `UseCython.cmake` module for this purpose](https://github.com/scikit-build/scikit-build/blob/main/skbuild/resources/cmake/UseCython.cmake). However, the successor package [`scikit-build-core`](https://github.com/scikit-build/scikit-build-core) does not. I originally considered porting the functionality there, but the Cython project itself seems like a more appropriate home for this logic. Moreover, putting this code into Cython would allow other builders to also take advantage of it instead of relying on `scikit-build-core`.

### Additional context

The one caveat with putting this module into Cython is that if Cython is installed via pip the module will go into the Python site-packages prefix. This location will not be on CMake's module search path by default, unlike if it was installed directly into the primary $PREFIX directly (e.g. a system /usr/lib directory equivalent, or perhaps a conda env prefix). That means that users would either need to query Python's CLI using a CMake custom command to get the site directory, or CMake would need to add support for this directly. Some discussion of that is in [the CMake issue I originally opened regarding Cython support](https://gitlab.kitware.com/cmake/cmake/-/issues/24982).

**Comment by henryiii** [comment]
FYI, a good example of supporting this, without requiring CMake, is at https://github.com/wjakob/nanobind/blob/master/cmake/nanobind-config.cmake. By supplying a cmake/cython-config.cmake file in the PyPI package, `find_package(cython)` will work out-of-the-box in scikit-build-core. (As mentioned above, there's a bit of querying that you'd have to do if you are using CMake directly; that can be simplified a little by providing a helper function that produces the location of the cmake module; and I'm hoping to improve support for this in CMake in the future.

**Comment by h-vetinari** [comment]
Meson also added native support for Cython: https://mesonbuild.com/Cython.html

**Comment by henryiii** [comment]
> How does the nanobind example avoid CMake? 

I might not have been clear - what I meant is `nanobind` itself doesn't use CMake to make SDists and wheels of nanobind. This is not like pybind11, which actually runs CMake when making an SDist in order to produce a config file.

If Cython added a cmake file similar to nanobind's, it would literally just be adding one file (plus any config necessary to make sure it's including in SDist/wheel), and it would just start working with scikit-build-core (and I think py-build-cmake too). The discussion in the linked issue is about how you'd also use it from pure CMake too, which should be pretty easy.

**Comment by vyasr** [comment]
Ah OK then yes, I agree. I've already started drafting a sample cython-config.cmake module for my own use that would essentially expose the `cython` CLI invocation via CMake. Happy to contribute that either here or in CMake proper if that is determined to be the correct home (although based on current discussions I think Cython is a better home for it). Then it's just a matter of adding an example of how to add the necessary site-packages dir to the search path using vanilla CMake as you mentioned, Marc showed us the necessary CMake vars in the other discussion.

**Comment by eli-schwartz** [comment]
What is the advantage of copying the `cython` CLI as properties compared to, say, a `cython_args` array of arguments to invoke cython with?

**Comment by eli-schwartz** [comment]
But don't all of those pros also apply to `*.cpp` files where you can specify `target_compile_options(target PRIVATE -o foobar.cpp.o)`?

But no one actually does that. They are far more likely to specify the std, in which case... last wins.

Although cmake does have a special property for defines, and also properties for things like -fvisibility=hidden or -fvisibility-inlines-hidden so presumably it is *sometimes* worth it...

...

The reason I bring this up is because meson (which I'm a developer for) has first-class builtin support for cython, and we offer 3 basic tunables for this:
- `cython_language={c|cpp}`
- `cython_version={2|3}`
- `cython_args=<array-of-string>`

We do not technically need the second option at all, but originally we unconditionally specified -3, and it's a wise default so we don't want to force everyone to set it in cython_args. The first option is crucial to select the output filename. Nothing else is really our responsibility, so we felt little motivation to offer distinctive target options, especially because it's important to ultimately give users the escape hatch of providing their own ~~CXXFLAGS~~ cython flags without updating their meson version to track updates to cython.

**Comment by vyasr** [comment]
> But don't all of those pros also apply to *.cpp files where you can specify target_compile_options(target PRIVATE -o foobar.cpp.o)?
>
> But no one actually does that. They are far more likely to specify the std, in which case... last wins.

I'm not sure I understand what you mean here. Are you saying that CMake allows arbitrary e.g. `CMAKE_CXX_FLAGS ` rather than providing a direct mapping of all possible options? If so, I agree and that's a good argument for why we should avoid the mapping. I'm not sure how that's related to "specifying the std" though, are you referring to flags like `-std=c++17`?

I'd be fine with stripping it down to the bare minimum of the language and an arbitrary set of `cython_args` like you suggest. The Python version (2 vs. 3) is also a reasonable default to include.

**Comment by eli-schwartz** [comment]
> I'm not sure how that's related to "specifying the std" though, are you referring to flags like `-std=c++17`?

Yes, which has a soft conflict with the use of CXX_STANDARD. That particular option (and a few others) could be specified either via a cmake property or via an arbitrary flag -- what happens if a project does both? Presumably, you get both.

**Comment by robertmaynard** [comment]
> But don't all of those pros also apply to `*.cpp` files where you can specify `target_compile_options(target PRIVATE -o foobar.cpp.o)`?

That won't break CMake :) . The template layout for the compilation line that CMake uses places the CMake computed `-o` as the last item on the compilation line. 

> Yes, which has a soft conflict with the use of CXX_STANDARD. That particular option (and a few others) could be specified either via a cmake property or via an arbitrary flag -- what happens if a project does both? Presumably, you get both, and then the compiler just uses whichever one comes last.

The very rough order of template line for C++ compilation in CMake in left to right:

- `CMAKE_CXX_FLAGS`
- computed target properties e.g. CXX_STANDARD
- user properties e.g. `target_compile_options`
- CMake's computed `-o <output_file>`

**Comment by eli-schwartz** [comment]
> That won't break CMake :) . The template layout for the compilation line that CMake uses places the CMake computed `-o` as the last item on the compilation line.

And the same rule can easily apply to cython integrations. ;) No matter how conceptually broken it is for a CMakeLists.txt to try doing that, required options can always be placed in the override position.

**Comment by eli-schwartz** [comment]
That looks like what I was thinking of (with the caveat that I'm not the world's biggest expert on cmake custom functions).

I agree that consuming the CMAKE_* namespace would be out of scope for something not shipping with upstream cmake, yeah.

**Comment by vyasr** [comment]
Cool thanks. We'll see what @henryiii has to say, if he likes this then I think it'll be in good shape for the Cython devs to weigh in.

**Comment by henryiii** [comment]
On flight to SciPy, so might be a bit delayed looking at this. If anyone is at SciPy we can discuss there a bit.

**Comment by vyasr** [comment]
Thanks for the heads up. No rush, just ping here whenever you have the chance to look. I am unfortunately not at SciPy this year, maybe next year!

**Comment by jcfr** [comment]
@henryiii, @thewtex and myself are now attending the SciPy sprints and will be working on this :rocket: 

We will consolidate existing work and propose a pull request to cython.

Depending on our finding, we may also create a `scikit-build/cython-cmake` project allowing us to (1) decouple scikit-build release from update to this new module and (2) deprecate the package to promote the use of the one that will be available in cython proper.

**Comment by vyasr** [comment]
Sounds great! Please keep me in the loop as well. If my existing prototype is close enough to what you envision, I'm happy to do the work myself and save you the trouble of recreating the wheel. I'm on board with the scikit-build -> Cython gradual progression and can push changes to either or both as you see fit.

**Comment by jcfr** [comment]
>  I'll be happy to include it in 3.0.1 or whenever you deem it ready.

Thanks for accommodating, we will follow up shortly.

**Comment by vyasr** [comment]
Please assign this issue to me so that I don't lose track of it. Thanks!

**Comment by henryiii** [comment]
Is there some way to detect the target language of a module before compiling it? For a CMake helper, it would be nice to know what language is being targeted, since it affects the output filename, and we need to tell CMake what file it is going to generate.

**Comment by da-woods** [comment]
@henryiii 

The normal way to indicate the language if it's not specified by the build system is to use `# distutils: language = ...` as a comment.

In Cython/Build/Dependencies.py the class `DistutilsInfo` handles it - it has a very simple and cut-down parser to identify this type of directive comment. I don't think there's a standalone interface that uses that right now, but that's definitely where I'd start from.

**Comment by henryiii** [comment]
Why would you pass multiple files to cythonize? If you pass a `.pyx` file it creates a `.c/.cxx` file (that you can specify with `--output-file`), but why would you pass multiple files? Trying to figure out if the cmake function should also support multiple files and how.

**Comment by da-woods** [comment]
If you call `cython.py` or `cythonize.py` with multiple `.pyx` files it looks like it just loops through them and generates multiple `.c` or `.so` files (each with an auto-generated name). As far as I can tell, the only reason to do that is to save a bit of typing on the command line.

-----------------------------------------

In the setuptools/distutils world you can do:

```
# a.pyx
cdef extern int some_cfunc()

def f():
    return some_cfunc()
```

```
# b.pyx
cdef public int some_cfunc():
    return 1
```

```
# setup.py
from setuptools import Extension, setup
from Cython.Build import cythonize

extensions = [
    Extension("a", ["a.pyx", "b.pyx"]),
]
setup(
    ext_modules=cythonize(extensions),
)
```

That links together multiple files into one extension and generates a warning:

```
Warning: Multiple cython sources found for extension 'a': ['a.pyx', 'b.pyx']
See http://cython.readthedocs.io/en/latest/src/userguide/sharing_declarations.html for sharing declarations among Cython files.
```

I think this is mainly a historical artefact from before `pxd` files so there's no reason for a new build system to support it.

**Comment by henryiii** [comment]
Okay, here's my proposed helper function:

```cmake
cythonize(<pyx_file>
         [LANGUAGE C | CXX]
         [CYTHON_ARGS <args> ...]
         [OUTPUT <OutputFile>]
         [OUTPUT_VARIABLE <OutputVariable>]
         )
```

If LANGUAGE is missing, it tries to deduce it. You can set the OUTPUT to a path. You can use OUTPUT_VARIABLE to get the path to the output file. I'm currently not supporting multiple filenames, as I think that would be problematic for OUTPUT (as I assume the matching cython feature is for `--output-file`) and OUTPUT_VARIABLE, which is a single file rather than a list. I've basically implemented this helper in https://github.com/scikit-build/cython-cmake/pull/19 but with a different name so that it won't clash with an upstream one, but it can provide a backport with the same name if it makes it in for older Cythons.

**Comment by vyasr** [comment]
I would avoid using `cythonize` as a name. That will be confusing because Cython has two CLIs, `cython` and `cythonize`, where the former only generates C/C++ code while the latter can also produce the resulting extension module, so I might expect a CMake function named `cythonize` to also handle the actual compilation. Aside from that, I think the main open question to me is whether the language should be inferred or not. I'd lean towards following the same inference rules as the `cython` CLI, which is C unless specified otherwise (either via the CLI or in the file itself via directives). Otherwise this looks good to me.

**Comment by da-woods** [comment]
> Aside from that, I think the main open question to me is whether the language should be inferred or not. I'd lean towards following the same inference rules as the `cython` CLI, which is C unless specified otherwise (either via the CLI or in the file itself via directives)

I think inferring it from the directives in the source file is pretty unambiguous and useful, so we should do that.

**Comment by vyasr** [comment]
I agree. In the case where no directives are specified `cython` will default to C, though; do we want the same default behavior in the CMake function, or can that error if no directive is specified and no language argument is passed?
